README.md
# RLassoCox
A reweighted Lasso-Cox model for survival prediction and biomarker discovery.

# Details
Package: RLassoCox

Type: Package

Title: A reweighted Lasso-Cox by integrating gene interaction information

Version: 0.99.0

Date: 2020-10-21

Authors@R: c(person(given = "Wei", family = "Liu", email = "freelw@qq.com", role = c("cre", "aut"),comment = c(ORCID = "0000-0002-5496-3641")))

Depends: R (>= 4.1),igraph, Matrix, survival, glmnet, stats

Description: RLassoCox is a package that implements the RLasso-Cox model proposed by Wei Liu. The RLasso-Cox model integrates gene interaction information into the Lasso-Cox model for accurate survival prediction and survival biomarker discovery. It is based on the hypothesis that topologically important genes in the gene interaction network tend to have stable expression changes. The RLasso-Cox model uses random walk to evaluate the topological weight of genes, and then highlights topologically important genes to improve the generalization ability of the Lasso-Cox model. The RLasso-Cox model has the advantage of identifying small gene sets with high prognostic performance on independent datasets, which may play an important role in identifying robust survival biomarkers for various cancer types.

License: GPL(>=2)

biocViews: Survival, Regression, GeneExpression, GenePrediction, Network

BugReports: https://github.com/weiliu123/RLassoCox

BiocType: Software

Suggests: knitr

VignetteBuilder: knitr

# Index of help topics:
RLassoCox Reweighted Lasso-Cox model

cvRLassoCox Cross-validation for the RLasso-Cox model

dGMMirGraph The KEGG network

g.HuRI.EntrezID The HuRI network

mRNA_matrix The expression data

predict.RLassoCox Make predictions from a RLasso-Cox model

predict.cvRLassoCox Make predictions from a cross-validated RLasso-Cox model

rw Directed Random Walk

survData Survival data

# Examples
data(dGMMirGraph)

data(mRNA_matrix)

data(survData)

trainSmpl.Idx <- sample(1:dim(mRNA_matrix)[1], floor(2/3*dim(mRNA_matrix)[1]))

testSmpl.Idx <- setdiff(1:dim(mRNA_matrix)[1], trainSmpl.Idx)

trainSmpl <- mRNA_matrix[trainSmpl.Idx ,]

testSmpl <- mRNA_matrix[testSmpl.Idx ,]

res <- RLassoCox(x=trainSmpl, y=survData[trainSmpl.Idx ,], globalGraph=dGMMirGraph) lp <- predict(object = res, newx = testSmpl)

cv.res <- cvRLassoCox(x=trainSmpl, y=survData[trainSmpl.Idx ,], globalGraph=dGMMirGraph, nfolds = 5) cv.lp <- predict(object = cv.res, newx = testSmpl, s = "lambda.min")


NEWS
Changes in version 0.99.0 (2020-10-21)
+ Submitted to Bioconductor


R/RLassoCox.R
RLassoCox <- function(x, y, globalGraph=NULL, Gamma=0.3, DEBUG=TRUE, 
                        standardize=TRUE,  ...){
    if(standardize){
        x <- scale(x, center = TRUE, scale = TRUE)
    }
    commonmRNA <- intersect(colnames(x),V(globalGraph)$name)
    x <- x[, commonmRNA]
    # calculate the p-value of each gene using Cox PH model
    if(DEBUG) cat('Calculating Cox p-value...')  
    colnames(y) <- c("status", "time")
    Survdata.mRNA <- data.frame(x, y,check.names = TRUE)
    geneCoxZP <- matrix(NA,nrow=ncol(x),ncol=2)
    rownames(geneCoxZP) <- colnames(x)
    colnames(geneCoxZP) <- c("ZScore","p-value")
    for(i in seq(from = 1, to = ncol(x))){
        res.coxph <- coxph(as.formula(paste("Surv(time, status)~", 
                        colnames(Survdata.mRNA)[i])), Survdata.mRNA)
        geneCoxZP[i,] <- summary(res.coxph)$coefficients[c(4,5)]
    }
    if(DEBUG) cat('Done\n')
    if(DEBUG) cat('Performing directed random walk...')
    p0 <- rep(0, length = vcount(globalGraph))
    names(p0) <- V(globalGraph)$name
    p0[rownames(geneCoxZP)] <- -log(geneCoxZP[, "p-value"])
    g.adj <- as_adjacency_matrix(globalGraph, type = "both", sparse = FALSE)
    PT <- rw(W = g.adj, p0 = p0, gamma = Gamma) 
    rownames(PT) <- names(p0)
    PT <- PT[commonmRNA,]
    if(DEBUG) cat('Done\n')
    selGene <- rownames(geneCoxZP)[which(geneCoxZP[, "p-value"] < 0.05)]
    PT <- PT[selGene]
    x <- x[, selGene]
    TPWeight <- 1/order(PT)
    res <- glmnet(x, y = Surv(time = y[,"time"], event = y[,"status"]),
                family = "cox", alpha = 1,  # lasso
                penalty.factor = TPWeight, ...)
    mod <- list(glmnetRes = res, PT = PT)
    class(mod) <- "RLassoCox"
    return(mod)
}


R/cvRLassoCox.R
cvRLassoCox <- function(x, y, globalGraph=NULL, nfolds = 10, Gamma=0.3, 
                        DEBUG=TRUE, standardize=TRUE,  ...){
    if(standardize){
        x <- scale(x, center = TRUE, scale = TRUE)
    }
    commonmRNA <- intersect(colnames(x),V(globalGraph)$name)
    x <- x[, commonmRNA]  
    # calculate the p-value of each gene using Cox PH model
    if(DEBUG) cat('Calculating Cox p-value...')
    colnames(y) <- c("status", "time")
    Survdata.mRNA <- data.frame(x, y,check.names = TRUE)
    geneCoxZP <- matrix(NA,nrow=ncol(x),ncol=2)
    rownames(geneCoxZP) <- colnames(x)
    colnames(geneCoxZP) <- c("ZScore","p-value")
    for(i in seq(from = 1, to = ncol(x))){
        res.coxph <- coxph(as.formula(paste("Surv(time, status)~", 
                        colnames(Survdata.mRNA)[i])), Survdata.mRNA)
        geneCoxZP[i,] <- summary(res.coxph)$coefficients[c(4,5)]
    }
    if(DEBUG) cat('Done\n')
    if(DEBUG) cat('Performing directed random walk...')
    p0 <- rep(0, length = vcount(globalGraph))
    names(p0) <- V(globalGraph)$name
    p0[rownames(geneCoxZP)] <- -log(geneCoxZP[, "p-value"])  
    g.adj <- as_adjacency_matrix(globalGraph, type = "both", sparse = FALSE)
    PT <- rw(W = g.adj, p0 = p0, gamma = Gamma) 
    rownames(PT) <- names(p0)
    PT <- PT[commonmRNA,]
    if(DEBUG) cat('Done\n')
    selGene <- rownames(geneCoxZP)[which(geneCoxZP[, "p-value"] < 0.05)]
    PT <- PT[selGene]
    x <- x[, selGene]
    TPWeight <- 1/order(PT)
    if(DEBUG) cat('Performing cv.glmnet...')
    res <- cv.glmnet(x, y = Surv(time = y[,"time"], event = y[,"status"]),
                family = "cox", nfolds = nfolds, alpha = 1,  # lasso
                penalty.factor = TPWeight, ...)
    if(DEBUG) cat('Done\n') 
    mod <- list(glmnetRes = res, PT = PT)
    class(mod) <- "cvRLassoCox"
    return(mod)
}
